components/support/utils/src/main/java/mozilla/components/support/utils/Browsers.kt

```diff
@@ -346,10 +348,22 @@ class Browsers private constructor(
             } else {
                 PackageManager.MATCH_DEFAULT_ONLY
             }
-            return packageManager.queryIntentActivities(intent, flag)
-                    .orEmpty()
-                    .filter { it.activityInfo.exported && (includeThisApp ||
-                            it.activityInfo.packageName != context.packageName) }
+            @Suppress("TooGenericExceptionCaught")
+            // We need to catch most general exception because sometimes
+            // TransactionTooLargeException is not properly caught
+            val appList = try {
```

Comment:
Could this be caused by large `data://` URLs? Do we want to filter them out?

Reply:
Yes, the exception is thrown only when we search with large data. I do not know how we could get the apps one by one, even with `MATCH_DEFAULT` flag the exception is thrown.

Reply:
Maybe @pocmo was referring to filtering `data://` (ignoring) if it's the source of the issue, right?

Reply:
@mcarare are large `data://` URLs the source of the issue? If so I agree that we should ignore them, are either way we will not be able to pass this URLs to third downloader apps.

Reply:
Sorry, I misread the question. In the case that was tested the file was a "fake" file, the URI being `data:application/pdf,0,1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,28,29,30,31,32,33,34,35,36,37,.... bunch of consecutive numbers` not a "data://"  URL.

I don't think we should filter those because there might be valid URI that can be handled, but I agree we should maybe handle differently an exception being thrown instead of querying with Intent  `    data = null`.
Should we just return Fenix if `includeThisApp ` is `true`, or an empty list for `false`?

Reply:
The above URL is a [data](https://developer.mozilla.org/en-US/docs/Web/HTTP/Basics_of_HTTP/Data_URIs) URL, that means the file itself it's embed on the URL,  this is what is causing the `TransactionTooLargeException`. I think the best approach is to not pass data URLs [here](https://github.com/mozilla-mobile/android-components/blob/f31c7c9ea0a74bff84a482fdd2f2c573510e03a1/components/feature/downloads/src/main/java/mozilla/components/feature/downloads/DownloadsFeature.kt#L329), as either way we are not going to be able to pass thes type of URLs to any third party app,  we could have a condition to identify data URLs  (`isDataUri = url.startsWith("data:")`) and just return an empty list. What do you think?

Reply:
What I meant is that the test case was not a **properly** formatted data URL (did not have a closing tag, for example, and the numbers were random, not real encoded data). So I was just wondering if a properly formatted one, and smaller than one that can cause OOM when passing it, could be handled just fine with our current code and we should not filter all of them, just the one that throws `TransactionTooLargeException`.

Reply:
Thanks for the clarification. I'm not completely sure if `data://` are widely supported if Android as it's more like a web standard. I more incline to just let the current browser to handle it.

Reply:
I Will do just that then: remove the catch and filter data URLs (meaning just returning Fenix for these cases and not launching a query).  
We can revisit this if we still get any reports for `TransactionTooLargeException ` or users' reports about not having another option than Fenix.

Reply:
Thank you both for making things clearer! 💯

Reply:
@Amejia481 wouldn't the check for data URL be better to be done in `DownloadsFeature`? I am not sure if changing `findResolvers` will affect `resolveBrowsers`. And, also, I cannot return `ResolveInfo` for Fenix without a query, and in Fenix we already add Fenix to the list when we call `getDownloaderApps`.

Reply:
Having the check in `DownloadsFeature` is OK, I think is the only consumer that could pass a `data:` URLs :)